metrics/src/builder.rs

```diff
@@ -0,0 +1,77 @@
+use crate::Receiver;
+use std::error::Error;
+use std::fmt;
+use std::time::Duration;
+
+/// Errors during receiver creation.
+#[derive(Debug, Clone)]
+pub enum BuilderError {
+    /// Failed to spawn the upkeep thread.
+    ///
+    /// As histograms are windowed, reads and writes require getting the current time so they can
+    /// perform the required maintenance, or upkeep, on the internal structures to roll over old
+    /// buckets, etc.
+    ///
+    /// Acquiring the current time is fast compared to most operations, but is a significant
+    /// portion of the other time it takes to write to a histogram, which limits overall throughput
+    /// under high load.
+    ///
+    /// We spin up a background thread, or the "upkeep thread", which updates a global time source
+    /// that the read and write operations exclusively rely on.  While this source is not as
+    /// up-to-date as the real clock, it is much faster to access.
+    UpkeepFailure,
```

Comment:
we should add a 

```rust
#[doc(hidden)]
_NonExahustive
```

Reply:
What does this buy us?

Reply:
if someone decides to match the error enum and we add new variants it wont break their code. Pretty much it forces them to have a catch all arm.

Reply:
Done!